### PR TITLE
fix bug in training model by amp

### PR DESCRIPTION
**Pull request by xiaohu2015**
This PR:
- Moves model prediction inside the `autocast` context
- Performs gradient clipping properly when autocast is enabled.
- Stops calling optimizer step twice when autocast is enabled.

### CODE REVIEW

**Comment by datumbox** [comment]
@xiaohu2015 Thanks for the PR. If there is no issue related to describe the bug, please add relevant information on the PR description so that it's clear what was the previous issue, etc.

**Comment by xiaohu2015** [comment]
> @xiaohu2015 Thanks for the PR. If there is no issue related to describe the bug, please add relevant information on the PR description so that it's clear what was the previous issue, etc.
> 
> @prabhat00155 Could you please have a look as you've recently work on this at #4547?



> @xiaohu2015 Thanks for the PR. If there is no issue related to describe the bug, please add relevant information on the PR description so that it's clear what was the previous issue, etc.
> 
> @prabhat00155 Could you please have a look as you've recently work on this at #4547?

@prabhat00155 can you review the PR? I found the amp was not working, so I update the training code to fix bugs.

**Review by prabhat00155 — COMMENTED** [review]
Thanks @xiaohu2015! Could you please upload the logs before and after the changes, since this is not covered by our unit tests?

**Review by datumbox — CHANGES_REQUESTED** [review]
Marking this as "Request changes" to avoid accidental merges before we gather enough about the nature of the bug, the training logs and we do proper investigation on our side.

@xiaohu2015 You can unblock this by adding context info on this PR as discussed above. Thanks!

**Comment by xiaohu2015** [comment]
> Thanks @xiaohu2015! Could you please upload the logs before and after the changes, since this is not covered by our unit tests?

After changing the code, I trained ResNet50, the result is 75.7 (use amp) and 75.5 (not use amp).

File: references/classification/train.py [hunk 1]
```diff
@@ -30,23 +30,20 @@ def train_one_epoch(model, criterion, optimizer, data_loader, device, epoch, arg
     for i, (image, target) in enumerate(metric_logger.log_every(data_loader, args.print_freq, header)):
         start_time = time.time()
         image, target = image.to(device), target.to(device)
-        output = model(image)
+        with torch.cuda.amp.autocast(enabled=args.amp):
+            output = model(image)
+            loss = criterion(output, target)
```
Comment (datumbox) [inline references/classification/train.py:35]: Good idea. Making both the forward and the loss estimation to be within the `autocast` call makes sense. [See documentation](https://pytorch.org/docs/stable/notes/amp_examples.html#typical-mixed-precision-training).

File: references/classification/train.py [hunk 2]
```diff
@@ -30,23 +30,20 @@ def train_one_epoch(model, criterion, optimizer, data_loader, device, epoch, arg
...
             scaler.update()
         else:
-            loss = criterion(output, target)
             loss.backward()
-
-        if args.clip_grad_norm is not None:
-            nn.utils.clip_grad_norm_(utils.get_optimizer_params(optimizer), args.clip_grad_norm)
-
-        optimizer.step()
+            if args.clip_grad_norm is not None:
```
Comment (datumbox) [inline references/classification/train.py:44]: The gradient clipping feature is now only supported when amp is not used. We should support it in both cases.

Note that you will need to unscale the gradients and then clip their norm. There is an example on the [documentation](https://pytorch.org/docs/stable/notes/amp_examples.html#gradient-clipping) on how to do this.

File: references/classification/train.py [hunk 3]
```diff
@@ -30,23 +30,20 @@ def train_one_epoch(model, criterion, optimizer, data_loader, device, epoch, arg
...
             scaler.step(optimizer)
             scaler.update()
         else:
-            loss = criterion(output, target)
             loss.backward()
-
-        if args.clip_grad_norm is not None:
-            nn.utils.clip_grad_norm_(utils.get_optimizer_params(optimizer), args.clip_grad_norm)
-
-        optimizer.step()
```
Comment (datumbox) [inline references/classification/train.py:49]: This looks like a bug in the previous implementation. It seems to me that since `scaler.step(optimizer)` calls the `optimizer.step()` internally, we were stepping twice.

**Review by datumbox — COMMENTED** [review]
@xiaohu2015 Your PR contains some good corrections but it's still very thin on information on the bug itself.

> After changing the code, I trained ResNet50, the result is 75.7 (use amp) and 75.5 (not use amp).

This is not necessarily an indication of an improvement. Doing multiple runs with different seeds can lead to slightly different results every time due to the random initialization and random transforms applied to the data.

File: references/classification/train.py (hunk 1 above)
Comment (prabhat00155) [reply references/classification/train.py:35]: Yeah, I agree. I think I missed it.

File: references/classification/train.py [hunk 4]
```diff
@@ -30,23 +30,20 @@ def train_one_epoch(model, criterion, optimizer, data_loader, device, epoch, arg
...
+            if args.clip_grad_norm is not None:
+                nn.utils.clip_grad_norm_(utils.get_optimizer_params(optimizer), args.clip_grad_norm)
```
[the remark this answers is not in the document]
Comment (sallysyw) [reply references/classification/train.py:45]: Thanks for bringing this out - I was just referring to ClassyVision's [implementation](https://github.com/facebookresearch/ClassyVision/blob/main/classy_vision/tasks/classification_task.py#L1206-L1207) before.

Given that the [official documentation](https://pytorch.org/docs/stable/notes/amp_examples.html#gradient-accumulation) is using `model.parameters()`, I think we can switch to it.

[the remark this answers is not in the document]
Comment (datumbox) [reply references/classification/train.py:45]: Classy might had it like this to support learnable params on the loss (we don't have this on Vision). Another reason might be that it was convenient in terms of code structure. 

@mannatsingh Do you have any idea why it was used like that in Classy?

[the remark this answers is not in the document]
Comment (mannatsingh) [reply references/classification/train.py:45]: Yes, so the only important reason that I can think of is that Apex's AMP works on its own (different) parameters which are disconnected from the model in certain settings (like O2). If you used the other approach, you would not actually be clipping the gradients. I'm not sure if torchvision even support Apex AMP though!

Other situations are manageable, for instance, if you optimize the model and the loss, you just need to make sure to use both everywhere (it's slightly risky but not a blocker).

[the remark this answers is not in the document]
Comment (datumbox) [reply references/classification/train.py:45]: Thanks a lot @mannatsingh, this was very helpful. I think this means we can with `model.parameters()`.

**Review by datumbox — COMMENTED** [review]
@xiaohu2015 I've marked as resolved all the "FYI" comments above and left only those that need to be addressed to merge the PR.

Effectively the only thing required is to add support of gradient clipping when `amp` is active. I provided a reference from the documentation on how to do it. It's worth refactoring slightly the code to simplify according to the comments.

Please let me know if you plan to continue working on the PR. Thanks!

**Comment by xiaohu2015** [comment]
> @xiaohu2015 Your PR contains some good corrections but it's still very thin on information on the bug itself.
> 
> > After changing the code, I trained ResNet50, the result is 75.7 (use amp) and 75.5 (not use amp).
> 
> This is not necessarily an indication of an improvement. Doing multiple runs with different seeds can lead to slightly different results every time due to the random initialization and random transforms applied to the data.

yes, the experment is to check that the amp is working, not to prove the model trained with amp is better.

**Comment by xiaohu2015** [comment]
> @xiaohu2015 I've marked as resolved all the "FYI" comments above and left only those that need to be addressed to merge the PR.
> 
> Effectively the only thing required is to add support of gradient clipping when `amp` is active. I provided a reference from the documentation on how to do it. It's worth refactoring slightly the code to simplify according to the comments.
> 
> Please let me know if you plan to continue working on the PR. Thanks!

thanks, I do some modification with the help of the [document](https://pytorch.org/docs/stable/notes/amp_examples.html#gradient-clipping)

**Review by datumbox — APPROVED** [review]
@xiaohu2015 Thanks for the PR, LGTM!

Given that the code was modified heavily and it's not covered with tests, it would be good to do a run on our side to confirm that everything works as expected.

@prabhat00155 Let me know if you have the bandwidth for this.

@sallysyw Concerning the simplification discussed at https://github.com/pytorch/vision/pull/4874/files#r744539820, is this something you would be interested in doing or shall we create an issue about it?

**Comment by prabhat00155** [comment]
> Given that the code was modified heavily and it's not covered with tests, it would be good to do a run on our side to confirm that everything works as expected.
> 
> @prabhat00155 Let me know if you have the bandwidth for this.

Yeah sure, let me kick off a training run.

**Comment by prabhat00155** [comment]
This runs fine. Here is the output log:
[output_log.txt](https://github.com/pytorch/vision/files/7507569/output_log.txt)

**Review by prabhat00155 — APPROVED** [review]
